sageresearch-app-sdk/src/main/java/org/sagebionetworks/research/sageresearch/viewmodel/ScheduleRepository.kt

```diff
@@ -176,20 +226,206 @@ open class ScheduleRepository(context: Context) {
         }
     }
 
-    private fun cacheSchedules(activityListV4: ScheduledActivityListV4) {
-        syncingSubscription.add(Observable.just(scheduleDao)
+    /**
+     * When a schedule fails to update to Bridge for whatever reason, it will be marked in the db as so.
+     * This function queries the db for those schedules and re-attempts to update them on Bridge.
+     */
+    fun syncFailedSchedules() {
+        scheduleRepoSubscriptions.add(Observable.just(scheduleDao)
                 .subscribeOn(Schedulers.io())
                 .subscribe({ dao ->
-                    entityConverter.fromScheduledActivityListV4(activityListV4)?.let { list ->
-                        if (!reconcileWithDictionary()) {
-                            dao.upsert(list)
-                        }
+                    val schedules = dao.activitiesThatNeedSyncedToBridge()
+                    schedules.forEach {
+                        updateScheduleToBridge(it)
                     }
                 }, {
                     // TODO: mdephillips 9/4/18 how to do logger in kotlin?
                 }))
     }
 
+    /**
+     * Creates and associates the taskRunUuid and the schedule guid so that it can accessed later when complete
+     * @param schedule of the task
+     * @return the new associated uuid with this schedule
+     */
+    fun createScheduleTaskRunUuid(schedule: ScheduledActivityEntity): UUID {
+        val uuid = UUID.randomUUID()
+        scheduleTaskRunUuidMap[uuid] = schedule.guid
+        return uuid
+    }
+
+    /**
+     * @param schedule that we are going to update on bridge
+     *                 schedule can be null if the user does a task without internet
+     * @param taskResult the result of the user doing the schedule task
+     */
+    fun updateSchedule(taskResult: TaskResult, taskRunUuid: UUID) {
```

Comment:
@liujoshua you would call this function after the TaskResult is completed

Reply:
Cool, do you want me to